_posts/2017-11-17-crates-are-not-safe.md

```diff
@@ -0,0 +1,100 @@
+We find that all 21 crates have dependencies that use `unsafe`, effectively
+making their use of `#![forbid(unsafe_code)]` moot.
+
+Of course, it is entirely possible that all uses of `unsafe` in crates.io are
+valid, necessary, and actually safe. But validating this, particularly as crates
+are updated, for all dependencies would be a significant challenge. It seems
+that the Rust ecosystem needs a better mechanism for helping developers ensure
+that the code they are using is actually type-safe.
+
+## Usage of the Standard Library is Pervasive
+
+The other challenge for Tock, as an embedded operating system, is the prevalent
+use of the standard library on crates.io. Running on bare-metal, Tock cannot use
+`std::` and only has access to the libraries provided by `core::`. Any
+dependency we would want to use for Tock needs to include `#![no_std]` so that
+the compiler does not try to include the standard library. Again we survey all
+of the crates on crates.io and find that 93% (11488/12360) of crates use the
+standard library (i.e. do not have `#![no_std]`). When considering only crates
+with 10k or more downloads, the ratio does not change (882/944). However, when
+including required dependencies, the number increases to 97% (12023/12360).
+This makes using any crates from crates.io on an embedded platform nearly
+impossible.
+
+Because many of the standard library modules are simply the `core::` crate
+re-exported, not all of these crates actually use functionality not supported on
+embedded platforms. We could (and
+[have](https://github.com/andersk/enum_primitive-rs/pull/14)) issue pull
+requests to the crates which can simply use `core::`, but this is a very manual
+process. Also, there seems to be a sense that using `#![no_std]` is
+[harmful](https://stackoverflow.com/questions/35951024/stdopsadd-or-coreopsadd#comment59560275_35951141)
+as it makes it difficult for the crate to use non-core standard library features
+in the future.
+
+Cargo or the compiler could really help here. Since many features of the
+standard library are simply from the `core::` crate, the compiler should be able
+to make a transparent mapping. This would continue to make it easy for
+developers to create libraries (they would not have to worry about core vs.
+std), while also supporting more crates on embedded platforms.
+
+## Are There Any Crates Tock Can Use?
+
+From our analysis, we find that there are 109 crates on crates.io that do not
+use `unsafe`, declare `#![no_std]`, and only use dependencies with the same
+properties. Again looking at more popular crates with 10k or more downloads,
+there are only 13 crates that qualify.
+
+Currently, our approach is to avoid using any external crates. To continue to
+provide the guarantee that capsules can contain untrusted code we need some
+mechanism to ensure that capsule dependencies do not use `unsafe`, and there
+seems to be [little
+support](https://internals.rust-lang.org/t/pre-rfc-cargo-safety-rails/5535) to
```

Comment:
"there seems to be little support to adding that type of feature" feels too strong to me. I might just cut that second point out of the clause and not include the link again (as you have it above anyways).